recipes/stlab/all/conanfile.py

```diff
@@ -11,48 +11,196 @@ class Stlab(ConanFile):
+        self.options["boost"].without_type_erasure = True
+        self.options["boost"].without_wave = True
+
+    def _fix_boost_components(self):
+        if self.settings.os != "Macos": return
+        if self.settings.compiler != "apple-clang": return
+        if Version(self.settings.compiler.version) >= "12": return
+
+        #
+        # On Apple we have to force the usage of boost.variant, because Apple's implementation of C++17 is not complete.
+        #
+        self.output.info("Apple-Clang versions less than 12 do not correctly support std::optional or std::variant, so we will use boost::optional and boost::variant instead.")
+        self.options.boost_optional = True
+        self.options.boost_variant = True
+
+    def _default_task_system(self):
+        if self.settings.os == "Macos": 
+            return "libdispatch"
+
+        if self.settings.os == "Windows":
+            return "windows"
+
+        if self.settings.os == "Emscripten":
+            return "emscripten"
+
+        return "portable"
+
+    def _configure_task_system_auto(self):
+        self.options.task_system = self._default_task_system()
+
+    def _configure_task_system_libdispatch(self):
+        if self.settings.os == "Linux":
+            if self.settings.compiler != "clang":
+                self.options.task_system = self._default_task_system()
+                self.output.info("Libdispatch requires Clang compiler on Linux. The task system is changed to {}.".format(self.options.task_system))
+                return
+
+        elif self.settings.os != "Macos":
+            self.options.task_system = self._default_task_system()
+            self.output.info("Libdispatch is not supported on {}. The task system is changed to {}.".format(self.settings.os, self.options.task_system))
+            return
+
+    def _configure_task_system_windows(self):
+        if self.settings.os != "Windows":
+            self.options.task_system = self._default_task_system()
+            self.output.info("Libdispatch is not supported on {}. The task system is changed to {}.".format(self.settings.os, self.options.task_system))
+            return
+
+    def _configure_task_system_emscripten(self):
+        if self.settings.os != "Emscripten":
+            self.options.task_system = self._default_task_system()
+            self.output.info("Libdispatch is not supported on {}. The task system is changed to {}.".format(self.settings.os, self.options.task_system))
+            return
+
+    def _configure_task_system(self):
+        if self.options.task_system == "auto":
+            self._configure_task_system_auto()
+        elif self.options.task_system == "libdispatch":
+            self._configure_task_system_libdispatch()
+        elif self.options.task_system == "windows":
+            self._configure_task_system_windows()
+        elif self.options.task_system == "emscripten":
+            self._configure_task_system_emscripten()
```

Comment:
Perhaps set a good default in `config_options`, and checking whether the requested option is valid in `configure`?

Reply:
thanks for the suggestion.
the check should be in `validate` method, I guess. Am I wrong?

Reply:
@madebr you mean to remove the `auto` value from `"task_system": ["portable", "libdispatch", "emscripten", "pnacl", "windows", "auto"],`?
If so, how I avoid to rewrite the user selected option?
For example the user selects `-o task_system=libdispatch` on Linux, but the good default for that platform is `portable`.
Does `config_option()` executes before assigning the user option selection?

Reply:
@madebr @prince-chrismc guys, could you please give me more information on how to solve that? See the above questions. Thanks!

Reply:
Sorry for the delay this has been sitting in my inbox all week 🙊 

I think there's a lot of boilerplate that could be cut out and it's hard to see exactly what you were trying to do. 
That being said the general concept is in the right direction.

My only concern with your implementation is if the consumer **explicitly** sets and option... it would get overridden. This would be a nasty surprise if they missed the tiny info message. 

In these cases we know what the _correct_ answer is so we can fail the command... much more obvious.

***

What we typically do this in validate

```py
if self.settings.os == "Emscripten" and self.options.task_system != "emscripten":
    raise ConanInvalidConfiguration("{self.name}/{self.version} task_system MUST be emscripten when using OS: {self.settings.os}")
```

That would kill a lot of overriding... 👍 

As for the `"auto"` definitely keep it... just make it more obvious. at the top of configure

```py
if self.options.task_system == "auto":
            self.options.task_system = self._default_task_system() # Take which is best for the platform
```